### PR TITLE
Add text input component

### PR DESCRIPTION
**Pull request by mauritskorse**
# What does this implement/fix?

Adding a input text component (similar to [Home Assistant](https://www.home-assistant.io/integrations/input_text/)). It allows for setting a string by the user through webserver comp front-end (v1). Implementation for v2 webserver front-end and Home Assistant is not done yet.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [x] ESP8266

## Example entry for `config.yaml`:
```yaml
input_text:
  - platform: template
    name: "test input text"
    id: "it_1"
    icon: "mdi:text"
    optimistic: true
    internal: false

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by nagyrobi** [comment]
Would it be possible to implement `initial_value` and `restore_value` for this?

**Comment by mauritskorse** [comment]
> Would it be possible to implement `initial_value` and `restore_value` for this?

Yep, that's already on the roll.

**Comment by SCratORS** [comment]
I'm waiting for this component. Very lacking...

**Comment by nielsnl68** [comment]
Me too, and for sure the HA integration part.

**Comment by jesserockz** [comment]
This will not work with Home Assistant. 
Integrations cannot create "Input helpers". This needs to wait until there is a core entity type that allows text inputs, just like `number` did for `input_number`

**Comment by nielsnl68** [comment]
I hope this can be added soon. I like to use this in my switchplate component.

**Comment by mauritskorse** [comment]
I have tried to create a new text component for HA, but I can't get it to work yet. Not sure what I'm missing. Maybe someone else could have a try.
Here's the code: https://github.com/home-assistant/core/commit/736f01a70e80b79b87ee465d5509b13397f7686c

**Comment by mauritskorse** [comment]
> I have tried to create a new text component for HA, but I can't get it to work yet. Not sure what I'm missing. Maybe someone else could have a try. Here's the code: [home-assistant/core@736f01a](https://github.com/home-assistant/core/commit/736f01a70e80b79b87ee465d5509b13397f7686c)

Ah, I already see what I'm missing. Will need to add support for this in [esphome/aioesphomeapi](https://github.com/esphome/aioesphomeapi) as well.

**Comment by mauritskorse** [comment]
Ok, I have tried it but I run into a problem testing as I can't get my aioesphomeapi branch to be used as dependency for the esphome component in HA (https://discord.com/channels/330944238910963714/554842238073700352/1039928836994908203). Maybe someone else could have a look. Below are links to the relevant branches
- https://github.com/mauritskorse/homeassistant-core/tree/input-text
- https://github.com/mauritskorse/aioesphomeapi/tree/text-input-component

**Comment by SeByDocKy** [comment]
Will be ready for the new 2022.12.0 HA release ? :)

**Comment by mauritskorse** [comment]
> Will be ready for the new 2022.12.0 HA release ? :)

I see that in the HA 2022.12 the `text` component was added. I will have a look whether I can get it to work with ESPHome now.

File: esphome/components/template/text/__init__.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+def validate(config):
+    if CONF_LAMBDA in config:
+        if config[CONF_OPTIMISTIC]:
+            raise cv.Invalid("optimistic cannot be used with lambda")
+        if CONF_INITIAL_VALUE in config:
+            raise cv.Invalid("initial_value cannot be used with lambda")
+        if CONF_RESTORE_VALUE in config:
+            raise cv.Invalid("restore_value cannot be used with lambda")
+    elif CONF_INITIAL_VALUE not in config:
```
Comment (nielsnl68) [inline esphome/components/template/text/__init__.py:30]: Why not use the `default=""` in the CONFIG_SCHEMA?

File: esphome/components/template/text/__init__.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    text.TEXT_SCHEMA.extend(
+        {
+            cv.GenerateID(): cv.declare_id(TemplateText),
+            cv.Optional(CONF_MAX): cv.int_,
+            cv.Optional(CONF_MIN): cv.int_,
+            cv.Optional(CONF_PATTERN): cv.string,
+            cv.Optional(CONF_LAMBDA): cv.returning_lambda,
+            cv.Optional(CONF_OPTIMISTIC, default=False): cv.boolean,
+            cv.Optional(CONF_SET_ACTION): automation.validate_automation(single=True),
+            cv.Optional(CONF_INITIAL_VALUE): cv.string_strict,
```
Comment (nielsnl68) [inline esphome/components/template/text/__init__.py]:
```suggestion
            cv.Optional(CONF_INITIAL_VALUE, default=""): cv.string_strict,
```

File: esphome/components/template/text/__init__.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    CONF_RESTORE_VALUE,
+    CONF_MAX,
+    CONF_MIN,
+    CONF_PATTERN,
+)
+from .. import template_ns
+
+TemplateText = template_ns.class_("TemplateText", text.Text, cg.PollingComponent)
+
+CONF_SET_ACTION = "set_action"
```
Comment (nielsnl68) [inline esphome/components/template/text/__init__.py:19]: All actions within **ESPHome** start with `on_***` where the `***` is replaced with a description of the action.

from what i understand you could use something like: `CONF_ON_MESSAGE`, `CONF_ON_STATE` or `CONF_ON_VALUE`

File: esphome/components/template/text/template_text.h [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  void set_initial_value(const std::string &initial_value) { this->initial_value_ = initial_value; }
+  void set_restore_value(bool restore_value) { this->restore_value_ = restore_value; }
+
+ protected:
+  void control(const std::string &value) override;
+  bool optimistic_ = false;
+  std::string initial_value_;
+  bool restore_value_ = false;
+  Trigger<std::string> *set_trigger_ = new Trigger<std::string>();
+  optional<std::function<optional<std::string>()>> f_;
```
Comment (nielsnl68) [inline esphome/components/template/text/template_text.h:31]: in ESPHome it is required to use full name method and property names describing the use of it.

**Review by nielsnl68 — CHANGES_REQUESTED** [review]
I made some small comments and suggestion on finalizing the code.

Something i'm not clear of when i just see the yaml code; you are using `min:` and `max:` in it; but what will that tell?  is that the value that is only allowed a number between `min:` and `max:`, or does it tell the size of the allowed string?

**Comment by foxsam21** [comment]
I just found this PR and wanted to test it out unfortunately something does not work I'm using dev branch and this config:
```yaml
external_components:
  - source: github://pr#3638
    components: 
    - text

text:
  - platform: template
    name: "My Text Component"
    id: "my_text"
    optimistic: true
```
but when I try to run I get some errors:
```
ERROR Unable to load component text:
Traceback (most recent call last):
  File "<my_path>/esphome/lib/python3.9/site-packages/esphome/loader.py", line 162, in _lookup_module
    module = importlib.import_module(f"esphome.components.{domain}")
  File "/usr/lib/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 855, in exec_module
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "<my_config_path>/.esphome/external_components/b41ee16a/esphome/components/text/__init__.py", line 46, in <module>
    cv.OnlyWith(CONF_MQTT_ID, "mqtt"): cv.declare_id(mqtt.MQTTTextComponent),
AttributeError: module 'esphome.components.mqtt' has no attribute 'MQTTTextComponent'
```

**Comment by nagyrobi** [comment]
Would be nice if this would be interoperable with the upcoming `key_collector` component so that text inputs could be also used for entering PIN codes etc. just like matrix or wiegand keypads can be used.

https://github.com/esphome/esphome-docs/pull/2426
https://github.com/ssieb/custom_components/tree/master/components/key_collector

@ssieb

**Comment by mauritskorse** [comment]
> I made some small comments and suggestion on finalizing the code.
> 
> Something i'm not clear of when i just see the yaml code; you are using `min:` and `max:` in it; but what will that tell? is that the value that is only allowed a number between `min:` and `max:`, or does it tell the size of the allowed string?

Thanks @nielsnl68!
Considering the `min` and `max`, that is the way the HA text entity addresses these (https://www.home-assistant.io/integrations/input_text/#configuration-variables). I agree that `min_length` and `max_length` are actually better.

Considering your suggestions, I have tried to stay close to the way the other template implementations are done. For example in number template component
https://github.com/esphome/esphome/blob/6b7b0768755b21cca0e4b67515de93b9ca7b3821/esphome/components/template/number/__init__.py#L21

https://github.com/esphome/esphome/blob/6b7b0768755b21cca0e4b67515de93b9ca7b3821/esphome/components/template/number/__init__.py#L58

Of course I don't mind changing according to your suggestions, but my python skill aren't that good that I immediately know what to do...

**Comment by nielsnl68** [comment]
I still hope this will be added soon :D

**Comment by arthurkomatsu** [comment]
This PR looks great, what is the status of this PR? @jesserockz

File: esphome/components/api/api.proto
Comment (jesserockz) [inline esphome/components/api/api.proto:1]: The `id` in here needs to be updated to start at 97 please to come after the alarm_control_panel which has a PR ready to merge in https://github.com/esphome/aioesphomeapi

File: esphome/components/template/text/__init__.py (hunk 3 above)
Comment (jesserockz) [reply esphome/components/template/text/__init__.py:19]: `set_...` are used for template platforms which is called differently from `on_state` etc

File: esphome/components/template/text/template_text.cpp [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+  if (!this->restore_value_) {
+    value = this->initial_value_;
+    ESP_LOGD(TAG, "State from initial: %s", value.c_str());
+  } else {
+    this->pref_ = global_preferences->make_preference<uint32_t>(this->get_object_id_hash());
+    if (!this->pref_.load(&value)) {
+      value = this->initial_value_;
+    } else {
+      value = "";
+    }
```
Comment (jesserockz) [inline esphome/components/template/text/template_text.cpp:24]: `std::string` cannot actually be saved reliably inside the preferences as the length is not known/saved, so it cannot be loaded back out properly. We might just have to make it non-restorable for now.

File: esphome/components/template/text/template_text.h (hunk 4 above)
Comment (jesserockz) [reply esphome/components/template/text/template_text.h:31]: While this is true, this platform is a copy of all the other template platforms. We should go through and update all of them at the same time (individual PRs would be fine too per platform which are easier to merge =)

File: esphome/components/template/text/__init__.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+async def to_code(config):
+    var = cg.new_Pvariable(config[CONF_ID])
+    await cg.register_component(var, config)
+    await text.register_text(
+        var,
+        config,
+        #        min_length=config[CONF_MIN_LENGTH],
+        #        max_length=config[CONF_MAX_LENGTH],
+        #        pattern=config[CONF_PATTERN],
```
Comment (jesserockz) [inline esphome/components/template/text/__init__.py:66]: Why are these commented?

File: esphome/components/text/text_call.cpp
Comment (jesserockz) [inline esphome/components/text/text_call.cpp:1]: The Call should have a `validate` function that checks the incoming value meets the requirements of min/max and pattern

File: esphome/components/text/text_traits.h [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+enum TextMode : uint8_t {
+  TEXT_MODE_AUTO = 0,
+  TEXT_MODE_TEXT = 1,
+  TEXT_MODE_PASSWORD = 2,
+};
+
+class TextTraits {
+ public:
+  // Set/get the number value boundaries.
+  void set_min_length(int min_length) { min_length_ = min_length; }
```
Comment (jesserockz) [inline esphome/components/text/text_traits.h:19]: Please use `this->` when referring to member variables
```suggestion
  void set_min_length(int min_length) { this->min_length_ = min_length; }
```

File: esphome/components/web_server/web_server.cpp [hunk 8]
```diff
@@ -764,6 +791,50 @@ std::string WebServer::number_json(number::Number *obj, float value, JsonDetail
...
+  }
+  request->send(404);
+}
+
+std::string WebServer::text_json(text::Text *obj, const std::string &value, JsonDetail start_config) {
+  return json::build_json([obj, value, start_config](JsonObject root) {
+    set_json_id(root, obj, "text-" + obj->get_object_id(), start_config);
+    if (start_config == DETAIL_ALL) {
+      root["mode"] = (int) obj->traits.get_mode();
+    }
```
Comment (jesserockz) [inline esphome/components/web_server/web_server.cpp:831]: Should this not also include the min/max/pattern?

**Review by jesserockz — CHANGES_REQUESTED** [review]
Mostly looks good, a few comments below.

https://github.com/esphome/aioesphomeapi PR/branch is missing still.

**Comment by nielsnl68** [comment]
What is the status of this PR? Has the HA side still not been added?

**Comment by EternityForest** [comment]
I think I have everything for restore_value here:
https://github.com/EternityForest/esphome/tree/add-text-input-component

But after writing the code, I discovered I can't easily test with what I've got on hand, this branch does not compile for S2 boards anymore it seems, it just tries and fails to fetch an old patch3 toolchain package.

**Comment by EternityForest** [comment]
I think I have this synced back up with dev on this branch, with the exception that I'm not able to compile the protobufs using version 3.6.1, haven't had time to set up a VM for that, so this is compiled on 3.12.

Not sure if this work is useful, or if it would just take longer to review than to redo, but I figured I'd give it a
go, if nothing else for a chance to learn about protobufs.

https://github.com/EternityForest/aioesphomeapi/tree/text-input-component
https://github.com/EternityForest/esphome/tree/sync-add-text-input-component

**Pull request closed without merging**